text/0000-buildpack-registry-search-api.md

```diff
@@ -0,0 +1,232 @@
+# Meta
+[meta]: #meta
+- Name: Buildpack Registry Search API
+- Start Date: 2020-11-11
+- Author(s): @elbandito
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+This is a proposal for a service API that enables clients to search for buildpacks.
+
+# Definitions
+[definitions]: #definitions
+
+**API** - Application Programmer Interface
+
+**Buildpack Registry** - the CNB centeralized [registry](https://github.com/buildpacks/registry-index) that contains metadata for all the community supported buildpacks 
+
+**Service** - an independent web process running in the cloud that provides RESTful endpoints
+
+**Client** - represents a physical end-user or software interface e.g. CLI, web UI, etc.
+ 
+# Motivation
+[motivation]: #motivation
+
+As the CNB project makes progress towards supporting a centralized Buildpack Registry, I anticipate the need to quickly search for and identify specific buildpacks.  To support this, I propose the creation of an API search service that customers could use directly and/or indirectly via a web dashboard or CLI.
+
+- What use cases does it support?  
+  - Customers looking to quickly indentify usefulful buildpacks to build their applications.
+- What is the expected outcome?  
+  - Customers will be able to quickly find relevant buildpacks for their use cases.
+
+# What it is
+[what-it-is]: #what-it-is
+
+An external API service that exposes endpoints for retrieving buildpack metadata.  Initially, I envision this service providing a few GET endpoints to support plain text query searches, and individual buildpack data retrieval.  The API endpoints will be versioned, and follow OpenAPI and/or Json Schema/API standards.
+
+### Supported Endpoints
+
+- **GET /search?matches=text**
+
+  Retrievs all the buildpacks that satisfy the search query.  If the response contains more than 20 buildpacks, pagination will be used.  For example:
```

Comment:
1. I can't think of a single API that returns paging data only after X amount of results if that is what this statement is saying. I think paging data should always be return.
2. Can we get some details on what that would look like and the query parameters available to control paging?

Reply:
I think we should follow something like Github:

* A `page` param
* A `Link` header with format `<url>; rel="<type>"`

https://docs.github.com/en/free-pro-team@latest/rest/guides/traversing-with-pagination

Reply:
In the past, I've also used `limit` and `offset` params.  For example:
```
GET /items?limit=20&offset=20
```

But, I agree that the Github approach seems reasonable.  

Does this RFC need to commit to a specific pagination implementation?

Reply:
Re: pagination commitment

In my opinion it would be preferable because the spec that comes out of this would need to if we hope to make this available as an interface for other registries.

Reply:
I don't think we're going to spec this. Given that we have the `/v1/` version schema, it can definitely change in the future. The goal here should be to capture where we want to start. I'm fine with `limit` and `offset` too.

Reply:
> I don't think we're going to spec this. Given that we have the `/v1/` version schema, it can definitely change in the future. The goal here should be to capture where we want to start. I'm fine with `limit` and `offset` too.

Maybe I'm confused but I was under the impression that there would be a spec for the API similar to how we have a spec for the [registry storage](https://github.com/buildpacks/spec/blob/main/extensions/buildpack-registry.md#backing-storage). The purpose for the spec would be so that this API interface can be reimplemented by other non-official registries. Not spec'ing how pagination works would simply mean that the spec/interface is incomplete.

To be clear, I'm not proposing that this be clearly defined in the RFC (given I have no strong opinion on either choice) or that we have a spec before the implementation starts. But if our goal is what I understand it to be it would need to be spec'd.

Reply:
Not at this time. We're not seeking to support unofficial registries right now, and it's unclear if we ever would. The goal right now is to support the frontend for the official registry only (this is common for interfaces like npmjs.org and rubygems.org for example)